lib/Dialect/Ibis/Transforms/IbisPortrefLowering.cpp

```diff
@@ -0,0 +1,340 @@
+//===- IbisPortrefLowering.cpp - Implementation of PortrefLowering --------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+
+#include "circt/Dialect/Ibis/IbisDialect.h"
+#include "circt/Dialect/Ibis/IbisOps.h"
+#include "circt/Dialect/Ibis/IbisPasses.h"
+#include "circt/Dialect/Ibis/IbisTypes.h"
+
+#include "mlir/IR/Builders.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace ibis;
+
+namespace {
+
+template <typename TPort>
+class PortLoweringPattern : public OpConversionPattern<TPort> {
+public:
+  PortLoweringPattern(MLIRContext *context)
+      : OpConversionPattern<TPort>(context) {}
+};
+
+class InputPortConversionPattern : public PortLoweringPattern<InputPortOp> {
+public:
+  using PortLoweringPattern::PortLoweringPattern;
+  using OpAdaptor = typename PortLoweringPattern<InputPortOp>::OpAdaptor;
+
+  LogicalResult
+  matchAndRewrite(InputPortOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    PortRefType innerPortRefType = op.getType().cast<PortRefType>();
+    Type innerType = innerPortRefType.getPortType();
+    Direction d = innerPortRefType.getDirection();
+
+    // Canonical form check - canonicalization should have ensured that only
```

Comment:
Why not just run the canonicalizer on the op in question? Again, "Pass pipelines should not rely on the canonicalizer pass for correctness. They should work correctly with all instances of the canonicalization pass removed." (https://mlir.llvm.org/docs/Canonicalization/)

Reply:
Worked a bit with this, here's the issue; we don't want to run canonicalization, we want to run CSE. We do have a
> `void eliminateCommonSubExpressions(RewriterBase &rewriter, DominanceInfo &domInfo, Operation *op, bool *changed = nullptr);`

available. However, it crashes if used within one of our rewrite patterns (looks like it's trying to erase the same op multiple times, may flag it as an issue upstream).  
I don't think we're ever going to actually encounter this error anytime soon (this is an intermediate pass that runs directly after tunneling, and tunneling will not create ops that require canonicalization), hence I'd be fine with leaving it as is.

Reply:
So the message should probably be changed to indicate that one needs to run the cse pass, not the canonicalizer pass.